flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaDeserializationSchema.java

```diff
@@ -96,32 +142,107 @@ public void deserialize(ConsumerRecord<byte[], byte[]> record, Collector<RowData
+
+		private final List<RowData> buffer = new ArrayList<>();
+
+		@Override
+		public void collect(RowData record) {
+			buffer.add(record);
+		}
+
+		@Override
+		public void close() {
+			// nothing to do
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	/**
+	 * Emits a row with key, value, and metadata fields.
+	 *
+	 * <p>The collector is able to handle the following kinds of keys:
+	 * <ul>
+	 *     <li>No key is used.
+	 *     <li>A key is used.
+	 *     <li>The deserialization schema emits multiple keys.
+	 * </ul>
+	 */
+	private static final class OutputProjectionCollector implements Collector<RowData>, Serializable {
 
 		private static final long serialVersionUID = 1L;
 
+		private final int[] keyProjection;
+
+		private final int[] valueProjection;
+
 		private final MetadataConverter[] metadataConverters;
 
 		private transient ConsumerRecord<?, ?> inputRecord;
 
+		private transient List<RowData> physicalKeyRows;
+
 		private transient Collector<RowData> outputCollector;
 
-		MetadataAppendingCollector(MetadataConverter[] metadataConverters) {
+		OutputProjectionCollector(
+				int[] keyProjection,
+				int[] valueProjection,
+				MetadataConverter[] metadataConverters) {
+			this.keyProjection = keyProjection;
+			this.valueProjection = valueProjection;
 			this.metadataConverters = metadataConverters;
 		}
 
 		@Override
-		public void collect(RowData physicalRow) {
-			final GenericRowData genericPhysicalRow = (GenericRowData) physicalRow;
-			final int physicalArity = physicalRow.getArity();
+		public void collect(RowData physicalValueRow) {
+			// no key defined
+			if (keyProjection.length == 0) {
+				emitRow(null, (GenericRowData) physicalValueRow);
+				return;
+			}
+
+			// emit a value for each key
+			// if parsing key data gets problems, ignore the value data.
+			for (RowData physicalKeyRow : physicalKeyRows) {
+				if (physicalValueRow == null) {
+					physicalKeyRow.setRowKind(RowKind.DELETE);
+				}
+				emitRow((GenericRowData) physicalKeyRow, (GenericRowData) physicalValueRow);
+			}
+		}
+
+		@Override
+		public void close() {
+			// nothing to do
+		}
+
+		/**
+		 * There are 4 situations:
+		 * 1. key is null && value is not null => project and set rowkind = insert
+		 * 2. key is not null && value is not null => project and set rowkind = insert
+		 * 3. key is not null && value is null => project and set rowkind = delete
+		 * 4. key is null && value is null => impossible
+		 * This situation is impossible:
+		 *   keyProjection.length > 0 && key == null
+		 * */
+		private void emitRow(@Nullable GenericRowData physicalKeyRow, @Nullable GenericRowData physicalValueRow) {
 			final int metadataArity = metadataConverters.length;
+			final int physicalArity = keyProjection.length + valueProjection.length;
+			final RowKind rowkind = physicalValueRow == null ? RowKind.DELETE : physicalValueRow.getRowKind();
```

Comment:
Should also under `upsertMode==true`.

Reply:
It is a inner colletor. It will receive the tombstone only when the outer class `DynamicKafkaDeserializationSchema` decide to emitrow. Therefore, it doesn't need a flag to detect the mode of the deserialization.